pkg/output/output.go

```diff
@@ -53,64 +55,48 @@ func (o *Output) IsSuccessful() bool {
 	return true
 }
 
-func (o *Output) printFailures() {
-	if o != nil && len(o.failures) > 0 {
-		fmt.Println("printing out failures:")
-		for _, v := range o.failures {
-			fmt.Println(" - ", v)
-		}
+// Format can be used to retrieve the string for the output structure
+func (o *Output) Format(debug bool) string {
+	if o == nil {
+		return ""
 	}
-}
-
-func (o *Output) printExceptions() {
-	if o != nil && len(o.exceptions) > 0 {
-		fmt.Println("printing out exceptions preventing the verifier from running the specific test:")
-		for _, v := range o.exceptions {
-			fmt.Println(" - ", v)
-		}
+	output := ""
+	if debug {
+		output += format(o.debugLogs)
 	}
-}
-
-func (o *Output) printErrors() {
-	if o != nil && len(o.errors) > 0 {
-		fmt.Println("printing out errors faced during the execution:")
-		for _, v := range o.errors {
-			fmt.Println(" - ", v.Error())
-		}
+	if o.IsSuccessful() {
+		output += "All tests passed!\n"
+		return output
 	}
+	output += format(o.failures)
+	output += format(o.exceptions)
+	output += format(o.errors)
+	return output
 }
 
-func (o *Output) printDebugLogs() {
-	if o != nil && len(o.debugLogs) > 0 {
-		fmt.Println("printing out debug logs from the execution:")
-		for _, v := range o.debugLogs {
-			fmt.Println(" - ", v)
-		}
+func format[T any](slice []T) string {
+	if len(slice) == 0 {
+		return ""
 	}
+	output := ""
+	for _, value := range slice {
+		output += fmt.Sprintf(logFormat, value)
+	}
+	return output + "\n"
 }
 
 // Summary can be used for printing out output structure
 func (o *Output) Summary(debug bool) {
 	fmt.Println("Summary:")
-	if debug {
-		o.printDebugLogs()
-	}
-
-	if o.IsSuccessful() {
-		fmt.Println("All tests pass!")
-	} else {
-		o.printFailures()
-		o.printExceptions()
-		o.printErrors()
-	}
+	fmt.Print(o.Format(debug))
 }
 
 // Parse returns the data being stored on output
 // - failures as []error
 // - exceptions as []error
 // - errors as []error
-func (o *Output) Parse() ([]error, []error, []error) {
-	return o.failures, o.exceptions, o.errors
+func (o *Output) Parse() ([]error, []error, []error, []string) {
```

Comment:
I would rather that we create a new function to retrieve debug logs from this output struct rather than continue to extend this `Parse()` function.

However, maybe I don't understand the problem - what is missing that this PR hopes to achieve? The description isn't super clear

Reply:
Gotcha, yeah if `Format()` is all you need that would be great. If we find that we need `Parse()` I think we had better consider just making separate functions for each of:

```
	debugLogs []string
	// failures represents the failed validation tests
	failures []error
	// exceptions is to show edge cases where a verifier test couldn't be ran as expected
	exceptions []error
	// errors is collection of unhandled errors
	errors []error
```

or just making all of them public members in the struct

Reply:
Cool, removed Parse changes for now. We can handle it later if necessary